### ISSUE TITLE
Drop the obsolete numx backend

### ISSUE THREAD

**Issue by otizonaizit** [mdp-toolkit/mdp-toolkit#75]
Context: #58 

The `numx` backend can be dropped completely, and we can have `scipy` as a haard dependency.
After release 3.6 this is a low hanging fruit.

**Comment by Stewori** [comment]
I have second thoughts on removing numx, now that I realize it means scipy becoming a hard dependency. From my work on C-extension support for Jython I know that scipy stresses the Python C-"API" more than numpy and I expect scipy to be way more difficult to support for alternative Python implementations than CPython. E.g. I was able to get numpy working with Jython to some extend but not scipy. IMO there is nothing wrong with being able to use a fair subset of MDP in an environment that supports numpy but not scipy. Also PyPy is still struggling with C-extension support and this is still an active topic. E.g. at the recent language summit [HPy](https://github.com/pyhandle/hpy) was presented. You may argue non-Cpython support is a non-goal, but then I'd say keeping numx doesn't hurt and gives us more flexibility for the future to come. Who tells us that the numpy/scipy situation will stay like today if not even the CPython native API might?
Also, see the recent trouble caused by scipy 1.4: https://github.com/mdp-toolkit/mdp-toolkit/commit/fc771385774aece406df81c1403817d9b37e7ef5
(This fix would have still worked without numx but it supports my point that the greater flexibility provided by numx is valuable.)
Also the ability to experiment with numerical bindings beyond numpy and scipy is still attractive, e.g. consider [eigen](https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html).

**Comment by otizonaizit** [comment]
Re: non CPython support ⟶ I didn't think of non CPython support, you are right. On the other hand, I see MDP as part of the scientific Python ecosystem, in the sense that I expect MDP to be used along with other libraries in that ecosystem (matplotlib, jupyter, pandas, scikit-learn/-image), and pretty much everything depends on scipy. But sure, it makes sense to keep the subset of MDP which only depends on numpy to stay that way.

Regarding #69 , I think this is more of an argument to drop numx! The real fix would be to just use numpy everywhere, and use scipy only there where we need stuff that lives in scipy and not in numpy. I appreciate your fix, but I think that long term that is ugly, don't you think so?

So, what could be done would be to keep the numx thing, but remove scipy from the possible numx backends. Basically, as long as no one comes up with a replacement for numpy (eigen or something else), numx would point to numpy, and we would explicitly use scipy where it is needed, and treat scipy as an "extension" like we do with scikit-learn, joblib, etc.

**Comment by Stewori** [comment]
> that is ugly, don't you think so

Yes, it certainly is. I just went for the simplest and least invasive fix possible at the time. Switching numx->numpy everywhere where it's relevant would be tedious.

> numx would point to numpy, and we would explicitly use scipy where it is needed

Yes, let's do that. That's my favorite solution. For some numerical routines, scipy is faster, (I only know of svd though). Do you know others? We can check if scipy can be imported and in that case override known advantageous routines in numx with those of scipy by default.
